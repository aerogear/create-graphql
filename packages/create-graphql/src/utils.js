import shell from 'shelljs';
import chalk from 'chalk';
import ora from 'ora';
import spawn from 'cross-spawn-promise';
import updateNotifier from 'update-notifier';

import pkg from '../package.json';

const tic = chalk.green('✓');
const tac = chalk.red('✗');

const installYeoman = async () => {
  const spinner = ora('Installing Yeoman...');

  spinner.start();

  const command = 'npm';
  const args = ['install', '-g', 'yo'];
  const options = {
    shell: true,
    stdio: false,
  };

  try {
    await spawn(command, args, options);

    spinner.stop();

    console.log(`${tic} Yeoman installed!`);
  } catch (error) {
    spinner.stop();

    console.error(`${tac} There was an error while trying to install Yeoman:`, error);
  }
};

export const verifyYeoman = async () => { // eslint-disable-line import/prefer-default-export
  if (!shell.which('yo')) {
    console.error(`${tac} GraphQL CLI requires Yeoman to be installed.`);

    await installYeoman();
  }

  return true;
};

export const verifyVersion = async () => { // eslint-disable-line import/prefer-default-export
  return await updateNotifier({pkg}).notify();
}
