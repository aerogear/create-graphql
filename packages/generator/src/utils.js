import path from 'path';
import fs from 'fs';
import recast from 'recast';
import pkgDir from 'pkg-dir';

const { visit } = recast.types;

const rootPath = pkgDir.sync('.') || '.';

const getModelCode = modelPath => fs.readFileSync(modelPath, 'utf8');

const parseFieldToGraphQL = (field) => {
  const graphQLField = {
    name: field.name,
    description: field.description,
    required: !!field.required,
  };

  switch (field.type) {
    case 'Number':
      return {
        type: 'GraphQLInt',
        flowType: 'number',
        ...graphQLField,
      };
    case 'Boolean':
      return {
        type: 'GraphQLBoolean',
        flowType: 'boolean',
        ...graphQLField,
      };
    case 'ObjectId': {
      return {
        type: 'GraphQLID',
        flowType: 'string',
        ...graphQLField,
      };
    }
    default:
      return {
        type: 'GraphQLString',
        flowType: 'string',
        ...graphQLField,
      };
  }
};

const parseGraphQLSchema = (mongooseFields) => {
  const dependencies = [];
  const fields = Object.keys(mongooseFields).map((name) => {
    const field = parseFieldToGraphQL(mongooseFields[name]);

    if (dependencies.indexOf(field.type) === -1) {
      dependencies.push(field.type);
    }

    return field;
  });

  return {
    fields,
    dependencies,
  };
};

const getSchemaFieldsFromAst = (node) => {
  const astSchemaFields = node.arguments[0].properties;

  const fields = [];
  astSchemaFields.forEach((field) => {
    const name = field.key.name;
    const fieldDefinition = {};

    if (field.value.type === 'ArrayExpression') {
      return;
    }

    field.value.properties.forEach(({ key, value }) => {
      const fieldName = key.name;

      fieldDefinition[fieldName] = value.name || value.value;
    });

    fields[name] = {
      name,
      ...fieldDefinition,
    };
  });

  return fields;
};

const getSchemaDefinition = (modelCode) => {
  const ast = recast.parse(modelCode, {
    parser: {
      parse: source => require('babylon').parse(source, {
        sourceType: 'module',
        plugins: [
          'asyncFunctions',
          'asyncGenerators',
          'classConstructorCall',
          'classProperties',
          'flow',
          'objectRestSpread',
          'trailingFunctionCommas',
        ],
      }),
    },
  });

  let fields = null;

  visit(ast, {
    visitExpression: function visitExpression(expressionPath) { // eslint-disable-line object-shorthand
      const { node } = expressionPath;

      if (
        node.type === 'NewExpression' &&
        node.callee.object.name === 'mongoose' &&
        node.callee.property.name === 'Schema'
      ) {
        fields = getSchemaFieldsFromAst(node);

        this.abort();
      }

      return this.traverse(expressionPath);
    },
  });

  return parseGraphQLSchema(fields);
};

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === 'source') {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = () => {
  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const defaultConfig = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return {
      ...defaultConfig,
      ...customConfig,
    };
  } catch (err) {
    // Return the default config if the custom doesn't exist
    return defaultConfig;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory) => getCreateGraphQLConfig().directories[directory];

/**
 * Get the relative path directory between two directories specified on the config file
 * @param from {string} The calling directory of the script
 * @param to {[string]} The destination directories
 * @returns {string} The relative path, e.g. '../../src'
 */
export const getRelativeConfigDir = (from, to) => {
  const config = getCreateGraphQLConfig().directories;

  return to.reduce((directories, dir) => {
    const relativePath = path.relative(config[from], config[dir]);

    return {
      ...directories,
      [dir]: relativePath,
    };
  }, {});
};

export const getMongooseModelSchema = (model) => {
  const modelDir = getCreateGraphQLConfig().directories.model;

  const modelPath = path.resolve(`${modelDir}/${model}.js`);

  const modelCode = getModelCode(modelPath);

  return getSchemaDefinition(modelCode);
};

/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text) => {
  return text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
};

/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = text => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;