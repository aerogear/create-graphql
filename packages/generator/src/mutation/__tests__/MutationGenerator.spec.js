import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';
import fs from 'fs-extra';

import {
  getFileContent,
  getFixturePath,
} from '../../../test/helpers';

import { getConfigDir } from '../../utils';

const mutationGenerator = path.join(__dirname, '..');

it('generate mutation files', async () => {
  const folder = await helper.run(mutationGenerator)
    .withArguments('Example')
    .toPromise();

  const destinationDir = getConfigDir('mutation');
  const destinationTestDir = getConfigDir('mutation_test');

  assert.file([
    `${destinationDir}/ExampleAddMutation.js`, `${destinationDir}/ExampleEditMutation.js`,
  ]);

  const files = {
    add: getFileContent(`${folder}/${destinationDir}/ExampleAddMutation.js`),
    edit: getFileContent(`${folder}/${destinationDir}/ExampleEditMutation.js`),
    addTest: getFileContent(`${folder}/${destinationTestDir}/ExampleAddMutation.spec.js`),
    editTest: getFileContent(`${folder}/${destinationTestDir}/ExampleEditMutation.spec.js`),
  };

  expect(files).toMatchSnapshot();
});

it('generate mutation files with schema', async () => {
  const folder = await helper.run(mutationGenerator)
    .inTmpDir(dir =>
      fs.copySync(
        getFixturePath('Post'),
        path.join(dir, 'src/model/Post.js'),
      ),
    )
    .withArguments('Post Post')
    .toPromise();

  const destinationDir = getConfigDir('mutation');
  const destinationTestDir = getConfigDir('mutation_test');

  assert.file([
    `${destinationDir}/PostAddMutation.js`, `${destinationDir}/PostEditMutation.js`,
  ]);

  const files = {
    add: getFileContent(`${folder}/${destinationDir}/PostAddMutation.js`),
    edit: getFileContent(`${folder}/${destinationDir}/PostEditMutation.js`),
    addTest: getFileContent(`${folder}/${destinationTestDir}/PostAddMutation.spec.js`),
    editTest: getFileContent(`${folder}/${destinationTestDir}/PostEditMutation.spec.js`),
  };

  expect(files).toMatchSnapshot();
});
