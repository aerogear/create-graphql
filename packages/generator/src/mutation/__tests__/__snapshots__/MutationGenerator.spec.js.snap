exports[`test generate mutation files 1`] = `
Object {
  "add": "// @flow

import {
  GraphQLString,
  GraphQLNonNull,
} from \'graphql\';
import {
  mutationWithClientMutationId,
  toGlobalId,
} from \'graphql-relay\';

import Example from \'../model/Example\';

import ExampleLoader from \'../loader/ExampleLoader\';
import ExampleConnection from \'../connection/ExampleConnection\';

export default mutationWithClientMutationId({
  name: \'ExampleAdd\',
  inputFields: {
    example: {
      type: new GraphQLNonNull(GraphQLString),
      description: \'My example field\',
    },
  },
  mutateAndGetPayload: async ({ example }, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error(\'Unauthorized user\');
    }

    // TODO: mutation logic

    return {
      // id: id, // ID of the newly created row
      error: null,
    };
  },
  outputFields: {
    exampleEdge: {
      type: ExampleConnection.edgeType,
      resolve: async({ id }, args, { user }) => {
        // TODO: load new edge from loader

        const example = await ExampleLoader.load(
          user, id,
        );

        // Returns null if no node was loaded
        if (!example) {
          return null;
        }

        return {
          cursor: toGlobalId(\'example\', example),
          node: example,
        };
      },
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "addTest": "import { graphql } from \'graphql\';
import { schema } from \'../../schema\';
import { setupTest } from \'../../../test/helper\';

import User from \'../model/User\';
import Example from \'../model/Example\';

beforeEach(async () => await setupTest());

it(\'should not allow anonymous user\', async () => {
  const query = \`
    mutation M {
      ExampleAdd(input: {
        exampleField: \"Example field\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const rootValue = {};
  // No user should be passed to context since we are testing an anonymous session
  const context = {};

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});

it(\'should create a record on database\', async () => {
  const user = new User({
    name: \'user\',
    email: \'user@example.com\',
  });

  await user.save();

  const query = \`
    mutation M {
      ExampleAdd(input: {
        exampleField: \"Example field\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const rootValue = {};
  const context = { user };

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});
",
  "edit": "// @flow

import {
  GraphQLID,
  GraphQLString,
  GraphQLNonNull,
} from \'graphql\';
import {
  mutationWithClientMutationId,
  fromGlobalId,
} from \'graphql-relay\';

import Example from \'../model/Example\';

import ExampleType from \'../type/ExampleType\';
import ExampleLoader from \'../loader/ExampleLoader\';

export default mutationWithClientMutationId({
  name: \'ExampleEdit\',
  inputFields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
    example: {
      type: GraphQLString,
    },
  },
  mutateAndGetPayload: async (args, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error(\'Unauthorized user\');
    }

    const {
      id,
      example,
    } = args;

    // Check if the provided ID is valid
    const example = await Example.findOne({
      _id: fromGlobalId(id).id,
    });

    // If not, throw an error
    if (!example) {
      throw new Error(\'Invalid exampleId\');
    }

    // TODO: mutation logic

    // Clear dataloader cache
    ExampleLoader.clearCache(example._id);

    return {
      id: example._id,
      error: null,
    };
  },
  outputFields: {
    example: {
      type: ExampleType,
      resolve: (obj, args, { user }) => ExampleLoader.load(user, obj.id),
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "editTest": "import { graphql } from \'graphql\';
import { toGlobalId } from \'graphql-relay\';
import { schema } from \'../../schema\';
import { setupTest } from \'../../../test/helper\';

import User from \'../model/User\';
import Example from \'../model/Example\';

beforeEach(async () => await setupTest());

it(\'should not allow anonymous user\', async () => {
  const query = \`
    mutation M {
      ExampleEdit(input: {
        id: \"Example Id\"
        exampleField: \"Example field\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const rootValue = {};
  // No user should be passed to context since we are testing an anonymous session
  const context = {};

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});

it(\'should create a record on database\', async () => {
  const user = new User({
    name: \'user\',
    email: \'user@example.com\',
  });

  await user.save();

  const query = \`
    mutation M {
      ExampleEdit(input: {
        id: \"Example Id\"
        exampleField: \"Example field\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const rootValue = {};
  const context = { user };

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});
",
}
`;

exports[`test generate mutation files with Schema 1`] = `
Object {
  "add": "// @flow

import {
  GraphQLString,
} from \'graphql\';
import {
  mutationWithClientMutationId,
  toGlobalId,
} from \'graphql-relay\';

import Post from \'../model/Post\';

import PostLoader from \'../loader/PostLoader\';
import PostConnection from \'../connection/PostConnection\';

export default mutationWithClientMutationId({
  name: \'PostAdd\',
  inputFields: {
    title: {
      type: new GraphQLNonNull(GraphQLString),
    },
    user: {
      type: new GraphQLNonNull(GraphQLID),
    },
    slug: {
      type: GraphQLString,
    },
  },
  mutateAndGetPayload: async (args, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error(\'Unauthorized user\');
    }

    const {
      title,
      user,
      slug,
    } = args;

    // TODO: mutation logic

    return {
      // id: id, // ID of the newly created row
      error: null,
    };
  },
  outputFields: {
    postEdge: {
      type: PostConnection.edgeType,
      resolve: async({ id }, args, { user }) => {
        // TODO: load new edge from loader

        const post = await PostLoader.load(
          user, id
        );

        // Returns null if no node was loaded
        if (!post) {
          return null;
        }

        return {
          cursor: toGlobalId(\'post\', post),
          node: post,
        };
      },
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "addTest": "import { graphql } from \'graphql\';
import { schema } from \'../../schema\';
import { setupTest } from \'../../../test/helper\';

import User from \'../model/User\';
import Post from \'../model/Post\';

beforeEach(async () => await setupTest());

it(\'should not allow anonymous user\', async () => {
  const query = \`
    mutation M {
      PostAdd(input: {
        title: \"Example value\"
        user: \"Example value\"
        slug: \"Example value\"
      }) {
        postEdge {
          node {
            title
            user
            slug
          }
        }
      }
    }
  \`;

  const rootValue = {};
  // No user should be passed to context since we are testing an anonymous session
  const context = {};

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});

it(\'should create a record on database\', async () => {
  const user = new User({
    name: \'user\',
    email: \'user@example.com\',
  });

  await user.save();

  const query = \`
    mutation M {
      PostAdd(input: {
        title: \"Example value\"
        user: \"Example value\"
        slug: \"Example value\"
      }) {
        postEdge {
          node {
            title
            user
            slug
          }
        }
      }
    }
  \`;

  const rootValue = {};
  const context = { user };

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});
",
  "edit": "// @flow

import {
  GraphQLString,
} from \'graphql\';
import {
  mutationWithClientMutationId,
  fromGlobalId,
} from \'graphql-relay\';

import Post from \'../model/Post\';

import PostType from \'../type/PostType\';
import PostLoader from \'../loader/PostLoader\';

export default mutationWithClientMutationId({
  name: \'PostEdit\',
  inputFields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
    title: {
      type: new GraphQLNonNull(GraphQLString),
    },
    user: {
      type: new GraphQLNonNull(GraphQLID),
    },
    slug: {
      type: GraphQLString,
    },
  },
  mutateAndGetPayload: async (args, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error(\'Unauthorized user\');
    }

    const {
      id,
      title,
      user,
      slug,
    } = args;

    // Check if the provided ID is valid
    const post = await Post.findOne({
      _id: fromGlobalId(id).id,
    });

    // If not, throw an error
    if (!post) {
      throw new Error(\'Invalid postId\');
    }

    // TODO: mutation logic

    // Clear dataloader cache
    PostLoader.clearCache(post._id);

    return {
      id: post._id,
      error: null,
    };
  },
  outputFields: {
    post: {
      type: PostType,
      resolve: (obj, args, { user }) => PostLoader.load(user, obj.id),
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "editTest": "import { graphql } from \'graphql\';
import { toGlobalId } from \'graphql-relay\';
import { schema } from \'../../schema\';
import { setupTest } from \'../../../test/helper\';

import User from \'../model/User\';
import Post from \'../model/Post\';

beforeEach(async () => await setupTest());

it(\'should not allow anonymous user\', async () => {
  // TODO: specify fields to create a new Post
  const post = new Post({
    title: \'Example value\',
    user: \'Example value\',
    slug: \'Example value\',
  });

  await post.save();

  const postId = toGlobalId(\'Post\', post._id);

  const query = \`
    mutation M {
      PostEdit(input: {
        id: \"\${postId}\"
        example: \"Example Field to Update\"
      }) {
        post {
          title
          user
          slug
        }
      }
    }
  \`;

  const rootValue = {};
  // No user should be passed to context since we are testing an anonymous session
  const context = {};

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});

it(\'should edit a record on database\', async () => {
  const user = new User({
    name: \'user\',
    email: \'user@example.com\',
  });

  await user.save();

  // TODO: specify fields to create a new Post
  const post = new Post({
    title: \'Example value\',
    user: \'Example value\',
    slug: \'Example value\',
  });

  await post.save();

  const postId = toGlobalId(\'Post\', post._id);

  const query = \`
    mutation M {
      PostEdit(input: {
        id: \"\${postId}\"
        example: \"Example Field to Update\"
      }) {
        post {
          title
          user
          slug
        }
      }
    }
  \`;

  const rootValue = {};
  const context = { user };

  const result = await graphql(schema, query, rootValue, context);

  expect(result).toMatchSnapshot();
});
",
}
`;
